File: src/Sorting/QuickSort/quick_sort.py
from random import randint


def quick_sort(array):
    quick_sort_recursive(array, 0, len(array) - 1)


def quick_sort_recursive(array, p, r):
    if p < r:
        q = random_partition(array, p, r)
        quick_sort_recursive(array, p, q - 1)
        quick_sort_recursive(array, q + 1, r)


def random_partition(array, p, r):
    pivot = randint(p, r)
    array[pivot], array[r] = array[r], array[pivot]
    return partition(array, p, r)


def partition(array, p, r):
    x = array[r]
    i = p - 1

    for j in range(p, r):
        if array[j] <= x:
            i += 1
            array[i], array[j] = array[j], array[i]

    q = i + 1
    array[q], array[r] = array[r], array[q]
    return q

File: src/Sorting/QuickSort/QuickSort.ts
/**
 * Sorts a given array in place with quick sort algorithm.
 *
 * @param {number[]} array
 */
export function quickSort(array: number[]): void {
  quickSortRecursive(array, 0, array.length - 1);
}

/**
 * Recursive call to quick sort.
 *
 * @param {number[]} array
 * @param {number} low - index of first element
 * @param {number} high - index of last element
 */
function quickSortRecursive(array, low, high): void {
  if (low < high) {
    const pivot = randomPartition(array, low, high);
    quickSortRecursive(array, low, pivot - 1);
    quickSortRecursive(array, pivot + 1, high);
  }
}

/**
 * Chooses index used as pivot in partitioning
 *
 * @param {number[]} array
 * @param {number} p - index of first element
 * @param {number} r - index of last element
 */
function randomPartition(array, p, r): number {
  const random = Math.floor(Math.random() * (r - p)) + p;
  [array[random], array[r]] = [array[r], array[random]];
  return partition(array, p, r);
}

/**
 *
 * @param {number[]} array
 * @param {number} p - index of first element
 * @param {number} r - index of last element (pivot)
 *
 * @return {number} index of pivot element
 */
function partition(array: number[], p: number, r: number): number {
  let i = p;

  for (let j = p; j < r; j++) {
    if (array[j] <= array[r]) {
      [array[j], array[i]] = [array[i], array[j]];
      i++;
    }
  }

  [array[i], array[r]] = [array[r], array[i]];
  return i;
}
